### PR TITLE
Feature/solar orbiter

### PR DESCRIPTION
**Pull request by mlooz**
Added a user-defined problem inspired by the solar orbiter mission to the trajectory gym. The trajectory is modelled as a series of ballistic flybys with free parameters r_p and beta to influence the inclination in the final flyby.

### CODE REVIEW

File: pykep/trajopt/gym/_solar_orbiter.py [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+            x[-1] * self._seq[-1].radius,
+            x[-2],
+            self._seq[-1].mu_self,
+        )
+        a, e, i, W, w, E = ic2par(eph[0], v_out, self._common_mu)
+        final_perihelion = a * (1 - e)
+        # orbit should be as polar as possible, but we do not care about prograde/retrograde
+        corrected_inclination = abs(abs(i) % pi - pi / 2)
+
+        # check perihelion and ahelion bounds during the flight
```
Comment (wbnns) [inline pykep/trajopt/gym/_solar_orbiter.py:177]: s/ahelion/aphelion/

File: pykep/trajopt/gym/_solar_orbiter.py [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+        min_sun_distance = final_perihelion
+        max_sun_distance = AU
+
+        for l_i in range(self._n_legs):
+            # project lambert leg, compute perihelion and ahelion
```
Comment (wbnns) [inline pykep/trajopt/gym/_solar_orbiter.py:182]: s/ahelion/aphelion/

File: pykep/trajopt/gym/_solar_orbiter.py [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+        v_out = fb_prop(
+            l[-1].get_v2()[0],
+            v_P,
+            x[-1] * self._seq[-1].radius,
+            x[-2],
+            self._seq[-1].mu_self,
+        )
+        a, e, i, W, w, E = ic2par(r_P, v_out, self._common_mu)
+        print("Perihelion: ", (a * (1 - e)) / AU, " AU")
+        print("Ahelion: ", (a * (1 + e)) / AU, " AU")
```
Comment (wbnns) [inline pykep/trajopt/gym/_solar_orbiter.py:294]: s/Ahelion/Aphelion/

File: pykep/trajopt/gym/_solar_orbiter.py [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+        Vinf_launch = sqrt(Vinfx ** 2 + Vinfy ** 2 + Vinfz ** 2)
+        # We transform it (only the needed component) to an equatorial system rotating along x
+        # (this is an approximation, assuming vernal equinox is roughly x and the ecliptic plane is roughly xy)
+        earth_axis_inclination = 0.409072975
+        Vinfz = - Vinfy * sin(earth_axis_inclination) + Vinfz * cos(earth_axis_inclination)
+        # And we find the vinf declination (in degrees)
+        sindelta = Vinfz / Vinf_launch
+        declination = asin(sindelta) / np.pi * 180.0
+        # We now have the initial mass of the spacecraft
+        m_initial = launchers.atlas501(Vinf_launch / 1000.0, declination)
```
Comment (darioizzo) [inline pykep/trajopt/gym/_solar_orbiter.py:161]: I think they use a different launcher  Atlas V 411 maybe ESOC can give you the table vinf decl mass for this and we can add a launcher?

**Comment by darioizzo** [comment]
The code seems fine in terms of structure, I cannot review its correctness in terms od dynamics though.

- [] Add tests on specific chromosomes (see how its done for JUICE)
- [] Solve the problem (find a solution) and produce pretty and plots of the resulting trajectory in a notebook (we can use it for the docs as well as for me to be able to see if theres bug in the orbital mechanics part)
- [] See the comment son the launcher
- What happened to the CI? Can you have a deeper look?

**Comment by mlooz** [comment]
Thanks for the review! The CI didn't trigger at first because I had marked the PR as a draft. I agree that it is still not finished, but had to mark it as ready for review (and then push again) to trigger the CI.

File: pykep/trajopt/gym/_solar_orbiter.py (hunk 1 above)
Comment (mlooz) [reply pykep/trajopt/gym/_solar_orbiter.py]: Thanks for noticing!

Comment (wbnns) [reply pykep/trajopt/gym/_solar_orbiter.py]: 🙂 👍

File: pykep/trajopt/gym/_solar_orbiter.py (hunk 4 above)
Comment (mlooz) [reply pykep/trajopt/gym/_solar_orbiter.py]: ESOC gave me a table with vinf and mass, but not using declinations. Added it to the launchers.

**Comment by mlooz** [comment]
Can you take a look at this again? The solution in the example notebook reaches the desired inclination only halfway. Partly this is since it only considers the cruise phase and not the later Venus flybys that happen in resonance, partly because I limited the computation to one run with 3000 generations to keep it in reasonable time.

**Comment by darioizzo** [comment]
Why not adding also the resonant fly-bys? In general I would only add to pykep problems and notebook that are able to reproduce known trajectories or results ...

**Comment by darioizzo** [comment]
Any progress on this? Or plan?

**Comment by darioizzo** [comment]
is this ready to merge?

**Comment by wbnns** [comment]
@mlooz 👏 👏 👏 💪

**Comment by dietmarwo** [comment]
https://gist.github.com/dietmarwo/86f24e1b9a702e18615b767e226e883f list many good solutions for planet sequence
[earth, venus, venus, earth, venus, venus, venus, venus, venus, venus] and dummy_DSMs=True.
Many of these solutions are similar to the original mission (inclination > 33 deg and final perhelion = 0.28) which means that the 
UDP is probably correct. The plot function needs to be extended for dummy_DSMs=True.

**Comment by dietmarwo** [comment]
Updated solutions to compare all three solar orbiter models are at

https://gist.github.com/dietmarwo/86f24e1b9a702e18615b767e226e883f

solodsmtest.py - for pykep.trajopt.gym._solar_orbiter._solar_orbiter_udp_1dsm
solotest.py - for pykep.trajopt.gym._solar_orbiter._solar_orbiter_udp
solomgartest.py - for pykep.trajopt.gym._solar_orbiter._solo_mgar_udp

**Comment by darioizzo** [comment]
@mlooz shall I merge this?

**Comment by mlooz** [comment]
@darioizzo I've added the solutions found by @dietmarwo to the notebook. When the tests complete, this can be merged.

**Comment by wbnns** [comment]
Nice job @mlooz !!!!

**Comment by darioizzo** [comment]
I need to redo all the CI I think!!! :( move from travis to git action

**Comment by darioizzo** [comment]
Merging this, CI failiure due to old obsolete manylinux wheel. Need to fix this in a separate PR

**Pull request merged**